### PR TITLE
[Lang] Add more functions to math module

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

File: python/taichi/_funcs.py [hunk 1]
```diff
@@ -100,22 +100,39 @@ def _matrix_outer_product(self, other):
...
-    c = x * scale
-    s = y * scale
-    r = Matrix([[c, -s], [s, c]], dt=dt)
-    return r, r.transpose() @ A
+    U = Matrix.identity(dt, 2)
+    P = ops.cast(A, dt)
+    zero = ops.cast(0.0, dt)
+    # if A is the zero matrix we simply return the pair (I, A)
+    if (A[0, 0] == zero and A[0, 1] == zero and A[1, 0] == zero
+            and A[0, 0] == zero):
```
Comment (FantasyVR) [inline python/taichi/_funcs.py:118]: duplicate `A[0, 0] == zero`

**Review by FantasyVR — COMMENTED** [review]
Thanks!

File: python/taichi/_funcs.py (hunk 1 above)
Comment (neozhaoliang) [reply python/taichi/_funcs.py]: Oh, my mistake. It should be A[1, 1] == zero

File: python/taichi/_funcs.py [hunk 2]
```diff
@@ -100,22 +100,39 @@ def _matrix_outer_product(self, other):
...
-    x, y = A(0, 0) + A(1, 1), A(1, 0) - A(0, 1)
-    scale = (1.0 / ops.sqrt(x * x + y * y))
-    c = x * scale
-    s = y * scale
-    r = Matrix([[c, -s], [s, c]], dt=dt)
-    return r, r.transpose() @ A
+    U = Matrix.identity(dt, 2)
+    P = ops.cast(A, dt)
+    zero = ops.cast(0.0, dt)
+    # if A is the zero matrix we simply return the pair (I, A)
```
Comment (FantasyVR) [inline python/taichi/_funcs.py:116]:
```suggestion
    # if A is the zero matrix we simply return the pair (I, A)
```
```suggestion
    # if A is a zero matrix we simply return the pair (I, A)
```

**Review by FantasyVR — APPROVED** [review]
LGTM！ Just a few nits.

**Pull request merged**